### ISSUE TITLE
node ssr dont showing 404 page

### ISSUE THREAD

**Issue by polioan** [withastro/astro#5535]
### What version of `astro` are you using?

1.6.12

### Are you using an SSR adapter? If so, which one?

node

### What package manager are you using?

npm

### What operating system are you using?

Windows (but linux have same behavior)

### Describe the Bug

This behavior is similar on standalone and middleware mode. When using vite dev server - all ok, default or custom 404 page are showing. But when using build Instead of 404 page - plain text are showing.

run "npm run serve" for starting server in production (build) mode

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-cddzmd

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by wulinsheng123** [comment]
mark

**Comment by wulinsheng123** [comment]
![image](https://user-images.githubusercontent.com/27261522/209598254-6b2ddd33-33ac-4f4a-ac4f-cb91faad128a.png)

Does this page only appear in the dev?
Does It not appear in the Astro bundle?

**Comment by bluwy** [comment]
@wulinsheng123 Yes I think that's intentional for dev only if the user doesn't provide a custom 404 page. For production, it should default to what the server would respond for 404.

Looking into the issue, looks like this is because we don't pass `{ matchNotFound: true }` [here](https://github.com/withastro/astro/blob/0b50987584120e0c0e549f9ff838dc8879dbfa30/packages/integrations/node/src/middleware.ts#L13). But we can't pass it because it might be a middleware, other middlewares might want to handle the request for something else instead of 404.

I think the fix here is that, if we're building for `standalone` mode, we set `{ matchNotFound: true }`. For `middleware` mode, we can leave it as is, but a nice bonus feature is if the server build output export a second middleware that handles 404.html serving only.

**Comment by wulinsheng123** [comment]
Thanks for your tips.

**Comment by bluwy** [comment]
Fixed in #5701

**Comment by schroedan** [comment]
@bluwy Well, the problem was only half fixed. The 404 handler is missing in the middleware module exports. We can work around this in our case by building for `standalone` mode, setting `ASTRO_NODE_AUTOSTART` env to `disabled` and using the handler from the `standalone` module. But this is not a very clean way in my opinion.

**Comment by wulinsheng123** [comment]
> @bluwy Well, the problem was only half fixed. The 404 handler is missing in the middleware module exports. We can work around this in our case by building for `standalone` mode, setting `ASTRO_NODE_AUTOSTART` env to `disabled` and using the handler from the `standalone` module. But this is not a very clean way in my opinion.
could you give a minimal duplicated rep ?

**Comment by romannurik** [comment]
It's possible 298dbb89f2963a547370b6e65cafd2650fdb1b27 may have broken this again?

**Comment by lilnasy** [comment]
@romannurik Some of the issues introduced by #7754 (released in astro v2.9.7) were fixed recently.

Please check with the latest version and if it's still there check the [feat: 404](https://github.com/withastro/astro/issues?q=is%3Aissue+is%3Aopen+label%3A%22feat%3A+404%22) label to see if an issue for the bug is already open.

**Comment by romannurik** [comment]
> @romannurik Some of the issues introduced by #7754 (released in astro v2.9.7) were fixed recently.
> 
> Please check with the latest version and if it's still there check the [feat: 404](https://github.com/withastro/astro/issues?q=is%3Aissue+is%3Aopen+label%3A%22feat%3A+404%22) label to see if an issue for the bug is already open.

Thanks! Looks like I'm hitting this already reported issue, so I'll subscribe to updates there: #7881

**Comment by romannurik** [comment]
FWIW, this horrid hack *DOES* work for me:
```js
app.use(astroSsrHandler);
app.use((req, res, next) => {
  // only called if astroSsrHandler didn't handle this request originally
  req.url = '/404';
  astroSsrHandler(req, res, next);
});
```
So in my situation (see comment above), before this hack, visiting `/404`, or a page that returned `new Response(null, { status: 404 })` correctly rendered the 404, but visiting some random non-existent URL seemed to be skipped over by the middleware
